Show labels ui

aljesusg opened the pull request:
Fix https://github.com/kiali/kiali/issues/2547

## Workload List
![workloads](https://user-images.githubusercontent.com/3019213/77306803-76db6380-6cf8-11ea-9646-0718973c7f2e.png)

Now in workloads we remove the label validation and we add the warning in details
![missing](https://user-images.githubusercontent.com/3019213/77306799-7511a000-6cf8-11ea-8c57-40f1800012f4.png)

## App List
![apps](https://user-images.githubusercontent.com/3019213/77306808-7773fa00-6cf8-11ea-90b2-a135865bc20b.png)
## Service List
![services](https://user-images.githubusercontent.com/3019213/77306797-74790980-6cf8-11ea-8482-c3a8261c4ff2.png)
## when you Filter by a Label we light up the label applied
![lighted](https://user-images.githubusercontent.com/3019213/77306787-70e58280-6cf8-11ea-94b1-67facb4138b1.png) 
## Overview List
![Screenshot from 2020-03-23 11-12-44](https://user-images.githubusercontent.com/3019213/77306822-7c38ae00-6cf8-11ea-93fd-a018bfd6312d.png)

### Overview Card
![overview_card](https://user-images.githubusercontent.com/3019213/77306821-7ba01780-6cf8-11ea-8a8a-5fa2f32a27f7.png)

#### Overview Card Labels
![overview_card_labels](https://user-images.githubusercontent.com/3019213/77306812-78a52700-6cf8-11ea-8f51-b3b1764195b3.png)

jotak commented:
I think I need more context to understand this.

I have some microservices deployed, with labels. In workloads list, I see these labels (so far so good). It's quite interesting to be able to filter by label [1].

But in Applications list, I don't see any labels, why? "Applications" isn't a k8s object, it's an abstract view, so how am I supposed to define labels on Applications?

Similarly in Overview page I only see "0 labels" whereas I do have some labels defined on workloads, so I'm not sure why.

[1] A small note in filtering labels: labels are displayed as "key: value", but for filtering it needs an equal: "key=value", so it seems inconsistent

prachiyadav commented:
![Screenshot from 2020-03-24 23-40-49](https://user-images.githubusercontent.com/13483060/77461719-fcf6c780-6e28-11ea-96bc-79cd6d323429.png)
![Screenshot from 2020-03-24 23-40-42](https://user-images.githubusercontent.com/13483060/77461726-fff1b800-6e28-11ea-9b19-63787208a477.png)
![Screenshot from 2020-03-24 23-39-54](https://user-images.githubusercontent.com/13483060/77461731-0122e500-6e29-11ea-8f81-23117feb7ab3.png)
![Screenshot from 2020-03-24 23-39-23](https://user-images.githubusercontent.com/13483060/77461734-01bb7b80-6e29-11ea-8304-15edba7df871.png)
![Screenshot from 2020-03-24 23-39-01](https://user-images.githubusercontent.com/13483060/77461737-02eca880-6e29-11ea-8065-5b60e50abc9b.png)

jshaughn commented on src/pages/Overview/OverviewPage.tsx:
```diff
@@ -409,6 +414,34 @@ export class OverviewPage extends React.Component<OverviewProps, State> {
+  renderLabels(ns: NamespaceInfo): JSX.Element {
+    const button = (
+      <div style={{ color: ns.labels && PfColors.Blue400 }}>
+        {ns.labels ? Object.entries(ns.labels).length : 0} labels
```
I think it's worth handling the "1 label" case.

aljesusg replied:
If there is 1 label so ns labels will be true and then Object.entries(ns.labels).length will be 1 .

I don't understand your proposal @jshaughn

jshaughn replied:
Sorry, it's just that I don't want to see "1 labels".  I'd prefer to add a test for the singular case and display "1 label". So,
0 labels
1 label
N labels

aljesusg replied:
Ohhh understand sorry I didn't see it

aljesusg commented:
> I think I need more context to understand this.
> 
> I have some microservices deployed, with labels. In workloads list, I see these labels (so far so good). It's quite interesting to be able to filter by label [1].
> 
> But in Applications list, I don't see any labels, why? "Applications" isn't a k8s object, it's an abstract view, so how am I supposed to define labels on Applications?
> 
> Similarly in Overview page I only see "0 labels" whereas I do have some labels defined on workloads, so I'm not sure why.
> 
> [1] A small note in filtering labels: labels are displayed as "key: value", but for filtering it needs an equal: "key=value", so it seems inconsistent

For Applications I am showing all the workloads labels of that application. See for example bookinfo the application review as label version=v1,v2,v3

aljesusg commented:
@jotak @jshaughn 

Opened for repalce = by : https://github.com/kiali/kiali/issues/2550
Opened for wrap text https://github.com/kiali/kiali/issues/2558

The others comment s I'll address them here to be related with the visualization of labels.

Thanks

jotak commented:
> For Applications I am showing all the workloads labels of that application. See for example bookinfo the application review as label version=v1,v2,v3

Forget it, I wasn't using the latest backend code, that's why I didn't see them.
I would have some other questions or remarks but it's more about the whole feature & backend code than about this particular PR, so not going to create some noise here :)

aljesusg commented:
> @jotak @jshaughn
> 
> Opened for repalce = by : [kiali/kiali#2550](https://github.com/kiali/kiali/issues/2550)
> Opened for wrap text [kiali/kiali#2558](https://github.com/kiali/kiali/issues/2558)
> 
> The others comment s I'll address them here to be related with the visualization of labels.
> 
> Thanks

I'll do this issues in https://github.com/kiali/kiali/issues/2550

jotak commented on src/pages/Overview/Filters.ts:
```diff
@@ -52,7 +52,7 @@ export const labelFilter: FilterTypeWithFilter<NamespaceInfo> = {
   title: 'Label',
   placeholder: 'Filter by Label',
-  filterType: FilterTypes.custom,
+  filterType: FilterTypes.label,
```
As discussed  on IRC I think it would be better to have a different filter type here, something like "nsLabel", to not mix up with labels on workloads

jotak replied:
(since it seems the problem was already there before, I can open a separate ticket)

lucasponce commented:
In the overview page, it's visually inconsistency to use two different ways of tooltip:
![image](https://user-images.githubusercontent.com/1662329/77900354-393d8400-727e-11ea-9bcf-f3245a43de12.png)
![image](https://user-images.githubusercontent.com/1662329/77900401-42c6ec00-727e-11ea-810a-d027cabf78aa.png)

We need to unify that, @kiali/ux-reviewers could suggest, but I would unify it.

I'm +1 to align the labels with the header:
![image](https://user-images.githubusercontent.com/1662329/77900511-74d84e00-727e-11ea-9baf-ee8ab608f638.png)
This looks moved.

Also, in Service List I would move Details as last column, for consistency with Apps/Workloads:
![image](https://user-images.githubusercontent.com/1662329/77900746-cc76b980-727e-11ea-8a62-b266e3a66faa.png)

I know column was in this position before, but now I guess it makes sense.

I'm undecided if I like to change the "=" per ":", because to me these are two different things.
One is an operator in an expression (used also in other text inputs like find/hide in graph where "=" is used) and ":" is used to represent a value in an object.

Also Kubectl use "=" in all places (on label operator and in --show-labels) only uses ":" when using json object.

lwrigh commented:
I would suggest using the black tooltip in this case. I also want to take a closer look at the placement of the labels within the card when we do another pass at a card layout.

aljesusg commented:
## Next iteration let a list view in overview to display more information ( we will be able to add more information in the future with this )

![ezgif-4-8948fc97a164](https://user-images.githubusercontent.com/3019213/77926367-55551b80-72a6-11ea-9f72-d9d9f3a7a3d0.gif)


## Proposal

Change the buttons
![buttons](https://user-images.githubusercontent.com/3019213/77926407-600fb080-72a6-11ea-8597-c0bf51992e73.png)

by icons, for example in ManageIQ we used icons to show views
![Screenshot from 2020-03-30 16-52-13](https://user-images.githubusercontent.com/3019213/77926832-df9d7f80-72a6-11ea-9698-f77ffef6330a.png)

aljesusg commented:
@jotak @lucasponce I did all the changes proposed by UX. 

The problems related with replace = by : and others are in the fix PR.

Please review it

aljesusg commented:
![Screenshot from 2020-04-01 13-04-30](https://user-images.githubusercontent.com/3019213/78130306-59557a80-7419-11ea-8dec-bec234a0e401.png)
![Screenshot from 2020-04-01 13-04-08](https://user-images.githubusercontent.com/3019213/78130307-59ee1100-7419-11ea-9c66-d2fce5b64042.png)

lucasponce commented:
In the Overview page I don't know what to do with this:
![image](https://user-images.githubusercontent.com/1662329/78241306-75701f00-74e0-11ea-8750-dd6fca6c8eff.png)
![image](https://user-images.githubusercontent.com/1662329/78241347-8325a480-74e0-11ea-8990-f7c7124e928f.png)

If we split the Istio Config icon in a column, the mtls should have same consistency in the list view.

lucasponce commented:
I know these changes are in master and are not in the goal of the PR:
![image](https://user-images.githubusercontent.com/1662329/78241590-ed3e4980-74e0-11ea-8d6d-730a2a6123ab.png)
But as we are updating this toolbar, I'd appreciate if we can align the "Namespaces" with the filter and the "Show health for" with the control.

lucasponce commented:
Also, I'm 50/50 with this scenario:
![image](https://user-images.githubusercontent.com/1662329/78241873-5f169300-74e1-11ea-9237-1c318819bff7.png)

When showing namespaces in a card, the sorting control makes sense.

But when using the list, as a user it seems weird and I guess I'd expect to sort with the table headers:

![image](https://user-images.githubusercontent.com/1662329/78241897-6b9aeb80-74e1-11ea-9fbf-d8bba46f4182.png)

I'm undecided 50/50 on this, as pros: consistency within the page, cons: introduce a weird scenario or a table (when in other pages, tables are sorted by headers).

aljesusg commented:
> Also, I'm 50/50 with this scenario:
> ![image](https://user-images.githubusercontent.com/1662329/78241873-5f169300-74e1-11ea-9237-1c318819bff7.png)
> 
> When showing namespaces in a card, the sorting control makes sense.
> 
> But when using the list, as a user it seems weird and I guess I'd expect to sort with the table headers:
> 
> ![image](https://user-images.githubusercontent.com/1662329/78241897-6b9aeb80-74e1-11ea-9fbf-d8bba46f4182.png)
> 
> I'm undecided 50/50 on this, as pros: consistency within the page, cons: introduce a weird scenario or a table (when in other pages, tables are sorted by headers).

We can disable the sort when is a list, maybe we can ask this to UX

lucasponce commented:
> We can disable the sort when is a list, maybe we can ask this to UX

Yes, that was a possibility, to disable/hide it when it's a list and just enable the columns for sorting.
But probably UX should give an opinion, I'm 50/50, so both options work for me.

lwrigh commented:
Thought about this a bit more and it might make sense to keep the sort option for the card views since the namespace name is still present on the card and it would still make sense to be able to  sort that name by alphabetical or reverse alphabetical order.

aljesusg commented:
Ok I did the changes and is updated . Thanks @lwrigh @lucasponce 

- Fixed tag show Health in dropdown
- Now we use virtualList component to
- Vertical align in all cells ( include workloads, apps etc...)
- Hide sort dropdown and enabled sortable headers in list view
- Added TLS column in List View

![Screenshot from 2020-04-03 11-35-36](https://user-images.githubusercontent.com/3019213/78346436-55ebfb80-759f-11ea-889a-e19831c1724e.png)

![Screenshot from 2020-04-03 11-35-48](https://user-images.githubusercontent.com/3019213/78346434-54bace80-759f-11ea-8aad-9e27a4247155.png)

![Screenshot from 2020-04-03 11-35-59](https://user-images.githubusercontent.com/3019213/78346419-52587480-759f-11ea-9eb7-e1e60c0ba6bf.png)

lwrigh commented:
Looks good to me! Having the sortable headers enabled only for list view looks good to me, it looks way cleaner than having the sort field in the toolbar.

lucasponce commented:
Almost perfect, but one single detail:
![image](https://user-images.githubusercontent.com/1662329/78545561-43bac900-77fc-11ea-8d7e-5d220ee47f8f.png)

If we can align "Namespaces" title with the dropdown below on the left, it would be in sync with the rest of pages.

I.e. Graph:

![image](https://user-images.githubusercontent.com/1662329/78545671-75339480-77fc-11ea-976b-3fbc12e5834b.png)

Apps:

![image](https://user-images.githubusercontent.com/1662329/78545687-7c5aa280-77fc-11ea-81b1-bf5c39cd14ed.png)
